usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -9472,283 +9574,203 @@ static int read_adapter_config_file(STDLL_TokData_t * tokdata,
-                           " in config file '%s', expected domain number"
-                           " (2nd number)\n", __func__, fname);
-                break;
+
+            if (strcmp(c->key, "PKEY_MODE") == 0) {
+                rc = ep11_config_set_pkey_mode(ep11_data, fname, strval);
+                if (rc != 0)
+                    break;
+                continue;
             }
-            if (strncmp(token, "END", 3) == 0) {
-                TRACE_ERROR("%s Expected 2nd number, found '%s' in config "
-                            "file\n", __func__, token);
-                OCK_SYSLOG(LOG_ERR,
-                           "%s: Error: Expected valid domain"
-                           " number (2nd number), found '%s' in config file"
-                           " '%s'\n", __func__, token, fname);
-                rc = APQN_FILE_SYNTAX_ERROR_2;
-                break;
+
+            if (strcmp(c->key, "DIGEST_LIBICA") == 0) {
+                rc = ep11_config_set_libica(ep11_data, fname, strval);
+                if (rc != 0)
+                    break;
+                continue;
             }
-            if (check_n(&ep11_data->target_list, token, &apqn_i) < 0) {
-                OCK_SYSLOG(LOG_ERR, "%s: Error: Expected valid domain"
-                           " number (2nd number), found '%s' in config file"
-                           " '%s'\n", __func__, token, fname);
-                rc = APQN_FILE_SYNTAX_ERROR_3;
+        }
+
+        if (confignode_hastype(c, CT_NUMPAIRLIST)) {
+            list = confignode_to_numpairlist(c);
+
+            if ((strcmp(list->base.key, "APQN_WHITELIST") != 0 &&
+                 strcmp(list->base.key, "APQN_ALLOWLIST") != 0) ||
+                strcmp(list->end, "END") != 0) {
+                ep11_config_error_token(fname, list->base.key, list->base.line,
+                                        "APQN_ALLOWLIST ... END");
+                rc = CKR_FUNCTION_FAILED;
                 break;
             }
-            ep11_data->target_list.length++;
-            if (ep11_data->target_list.length > MAX_APQN) {
-                TRACE_ERROR("%s Too many APQNs in config file (max %d)\n",
-                            __func__, (int) MAX_APQN);
-                OCK_SYSLOG(LOG_ERR,
-                           "%s: Error: Too many APQNs in config file '%s'\n",
-                           __func__, fname);
-                rc = APQN_FILE_SYNTAX_ERROR_4;
-                break;
+
+            whitemode = 1;
+            confignode_foreach(e, list->value, k) {
+                if (!confignode_hastype(e, CT_NUMPAIR)) {
+                    ep11_config_error_token(fname, e->key, e->line,
+                                            "pair of NUMBERs");
+                    rc = CKR_FUNCTION_FAILED;
+                    break;
+                }
+
+                rc = ep11_config_add_apqn(ep11_data, confignode_to_numpair(e),
+                                          fname);
+                if (rc != 0)
```

Comment:
The routine returns a CK_RV, so we may check for rc != CKR_OK instead of rc != 0

Reply:
fixed (all similar cases)